### PR TITLE
feat(auth): distributed key verification

### PR DESCRIPTION
**Pull request by cairin**
## Changes proposed in this pull request

refactor(auth): remove the registry from the client service
```- Removes the idea of the central registry from the client service.
- Refactors display info url -> uri to match spec.
- Adds clientId to the jwk.
- Makes use of typeguards to verify jwk tokens.
```

feat(backend): create client keys services
```
Allows managing clients:
- Creating a client
- Adding keys to a client

Allows fetching a client key by kid.

Imports the auth package, to allow use of the JWK typing in the backend.
```

feat(backend): routing for client keys

feat(backend): add graphql admin endpoints for adding client and keys

## Context
Removes the need for the key directory, and does 
## Checklist

- [ ] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass

### CODE REVIEW

File: packages/auth/src/client/service.ts [hunk 1]
```diff
@@ -91,56 +89,47 @@ async function verifySig(
...
+    clientInfo.display.name !== clientDetails.name ||
+    clientInfo.display.uri !== clientDetails.uri ||
+    jwk.clientId !== clientDetails.id
+  )
+    return false
 
-  const { keys, ...registryClientInfo } = await getRegistryDataByKid(
-    deps,
-    jwk.kid
+  const currentKey = keys[0]
```
Comment (njlie) [inline packages/auth/src/client/service.ts:109]: The `kid` should refer to a specific key now, so it doesn't need to account for an array of them here.

**Comment by njlie** [comment]
Looks good so far, I just realize when reading this that I was unclear about the fact that the response from the key registry also has the new shape, which is essentially an inversion of what you've got for [ClientDetails](https://github.com/interledger/rafiki/blob/e8656fc2a2ff0b42cbc6aa85a1ad22a923f876a9/packages/auth/src/client/service.ts#L36) already.

```
{
  ...JWK stuff,
  client: {
    ...client stuff
  }
}

**Comment by cairin** [comment]
@njlie I've pushed the changes 👍  This simplifies the other side in the backend quite nicely. Just need to refactor..

File: packages/auth/src/client/service.ts [hunk 2]
```diff
@@ -1,14 +1,26 @@
 import * as crypto from 'crypto'
 import Axios from 'axios'
 import { importJWK, JWK } from 'jose'
-import { URL } from 'url'
 
 import { BaseService } from '../shared/baseService'
 import { IAppConfig } from '../config/app'
 
+export interface JWKWithRequired extends JWK {
+  // clientId is the custom field representing a client in the backend
```
Comment (wilsonianb) [inline packages/auth/src/client/service.ts:9]:
```suggestion
  // client.id is the custom field representing a client in the backend
```

File: packages/auth/src/client/service.ts [hunk 3]
```diff
@@ -91,56 +88,39 @@ async function verifySig(
...
-  if (!keys || !keys[0].kid || !keys[0].x) {
     return false
-  }
 
-  return !!(
-    verifyClientDisplay(clientInfo.display, registryClientInfo) &&
-    verifyJwk(jwk, keys[0])
-  )
+  if (key.exp && new Date() >= new Date(key.exp * 1000)) return false
+  if (key.nbf && new Date() < new Date(key.nbf * 1000)) return false
```
Comment (wilsonianb) [inline packages/auth/src/client/service.ts:114]: The previous validation check (`verifyJwk`) required `key.exp` and `key.nbf`. Is this more correct?

File: packages/backend/src/clientKeys/service.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    knex
+  }
+  return {
+    getKeyById: (keyId) => getKeyById(deps, keyId)
+  }
+}
+
+async function getKeyById(
+  deps: ServiceDependencies,
+  // In the form https://somedomain/keys/kid
```
Comment (wilsonianb) [inline packages/backend/src/clientKeys/service.ts:32]:
```suggestion
  // In the form https://somedomain/keys/{keyId}
```

File: packages/backend/src/clients/service.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    addKeyToClient: (options) => addKeyToClient(deps, options)
+  }
+}
+
+async function getClient(
+  deps: ServiceDependencies,
+  id: string
+): Promise<Client | undefined> {
+  const client = await Client.query(deps.knex).findById(id)
+  client.keys = await getClientKeys(deps, client.id)
```
Comment (wilsonianb) [inline packages/backend/src/clients/service.ts:52]: Does something like this work?
https://vincit.github.io/objection.js/api/query-builder/eager-methods.html#withgraphfetched
```suggestion
  const client = await Client.query(deps.knex).findById(id).withGraphFetched('keys')
```

Comment (cairin) [reply packages/backend/src/clients/service.ts:52]: It does. But I was trying to `orderBy` which was getting quite complicated to do with the relation mappings. And this is only used by the admin api so I thought it probably wasn't necessary.

File: packages/auth/src/client/service.ts (hunk 3 above)
Comment (cairin) [reply packages/auth/src/client/service.ts:114]: Neither `key.exp` or `key.nbf` are required to be defined (according to the type definition). So I shouldn't need to check that they both are. If they are required I can put them in the typeguard and change the typings.

All the tests passed, so I assumed it was working as @njlie intended  🤷

File: packages/backend/src/app.ts [hunk 6]
```diff
@@ -363,6 +372,19 @@ export class App {
...
               )
+            } else if (path === '/keys/{keyId}' && method === HttpMethod.GET) {
+              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+              // @ts-ignore
+              this.publicRouter.get(
+                toRouterPath(path),
+                createValidatorMiddleware<ClientKeysContext>(openApi, {
+                  path,
+                  method
+                }),
```
[the remark this answers is not in the document]
Comment (cairin) [reply packages/backend/src/app.ts:383]: So can I just rip it out for now?

File: packages/auth/src/client/service.ts (hunk 3 above)
Comment (njlie) [reply packages/auth/src/client/service.ts:114]: The type definition is correct, so this is more accurate to what I intended.

File: packages/backend/src/app.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/backend/src/app.ts:377]: Remove `createValidatorMiddleware` and add this to the router outside of these `for` loops

[the remark this answers is not in the document]
Comment (cairin) [reply packages/backend/src/app.ts]: @wilsonianb I have refactored this, please check it again 👍

File: packages/backend/src/graphql/schema.graphql [hunk 7]
```diff
@@ -479,6 +496,21 @@ input CreateAccountInput {
   publicName: String
 }
 
+type Client implements Model {
```
Comment (njlie) [inline packages/backend/src/graphql/schema.graphql:499]: One of the required fields in auth's `ClientDetails` type that isn't present here is `email`.

File: packages/backend/src/clientKeys/model.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { BaseModel } from '../shared/baseModel'
+
+import { JWKWithRequired } from 'auth'
+
+export class ClientKeys extends BaseModel {
+  public static get tableName(): string {
+    return 'clientKeys'
+  }
+
+  // The id should be the same as the kid.
```
Comment (njlie) [inline packages/backend/src/clientKeys/model.ts:10]: Is this still true as of the discussion from this morning? i.e. isn't the `id` the internal database id and the `kid` a fully-qualified URL that contains the `id`?

File: packages/backend/src/clientKeys/routes.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+export function createClientKeysRoutes(
+  deps: ServiceDependencies
+): ClientKeysRoutes {
+  return {
+    get: (ctx: ClientKeysContext) => getKeyById(deps, ctx)
+  }
+}
+
+// TODO spec
```
Comment (njlie) [inline packages/backend/src/clientKeys/routes.ts:20]: Can this be cleaned up or elaborated upon?

File: packages/backend/src/graphql/schema.graphql (hunk 7 above)
Comment (cairin) [reply packages/backend/src/graphql/schema.graphql:499]: And `image`. Will update it 👍

File: packages/backend/src/clientKeys/model.ts (hunk 8 above)
Comment (cairin) [reply packages/backend/src/clientKeys/model.ts:10]: Yes, will update the comment.

File: packages/backend/src/graphql/schema.graphql (hunk 7 above)
Comment (cairin) [reply packages/backend/src/graphql/schema.graphql:501]: This has been fixed.

**Review by wilsonianb — APPROVED** [review]

**Review by njlie — APPROVED** [review]

**Pull request merged**